src/hotspot/share/compiler/compilationPolicy.cpp

```diff
@@ -633,9 +633,9 @@ CompileTask* CompilationPolicy::select_task(CompileQueue* compile_queue) {
       task = next_task;
       continue;
     }
-    if (task->compile_reason() == CompileTask::Reason_Whitebox) {
-      // Whitebox (CTW) tasks do not participate in rate selection and/or any level
-      // adjustments. Just return them in order.
+    if (task->is_blocking() && task->compile_reason() == CompileTask::Reason_Whitebox) {
```

Comment:
This may be not enough - testing with `-Xcomp` sets `-Xbatch`: [arguments.cpp#L1358](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/arguments.cpp#L1358)

Reply:
On other hand, when running with `-Xcomp` we should not worry about priority of compilation tasks.
And `Level2RecompilationTest.java` "requires" `vm.compMode != "Xcomp"` so it should pass.

Reply:
Yes. Testing confirms both the affected test and CTW works fine.

I think the cleanest way is to tell `Whitebox` whether we are entering for CTW, or for faking a normal compilation. But it would significantly more intrusive.

Reply:
Right. I will test this change.